graal-js/src/com.oracle.truffle.js/src/com/oracle/truffle/js/builtins/TaintBuiltins.java

```diff
@@ -0,0 +1,142 @@
+        addTaint(0),
+        getTaint(0),
+        isTainted(0),
+        removeTaint(0);
+
+        private final int length;
+
+        Taint(int length) {
+            this.length = length;
+        }
+
+        @Override
+        public int getLength() {
+            return length;
+        }
+    }
+
+    @Override
+    protected Object createNode(JSContext context, JSBuiltin builtin, boolean construct, boolean newTarget, Taint builtinEnum) {
+        switch (builtinEnum) {
+            case addTaint:
+                return JSAddTaintNodeGen.create(context, builtin, args().fixedArgs(2).createArgumentNodes(context));
+            case getTaint:
+                return JSGetTaintNodeGen.create(context, builtin, args().fixedArgs(1).createArgumentNodes(context));
+            case isTainted:
+                return JSIsTaintedNodeGen.create(context, builtin, args().fixedArgs(1).createArgumentNodes(context));
+            case removeTaint:
+                return JSRemoveTaintNodeGen.create(context, builtin, args().fixedArgs(3).createArgumentNodes(context));
+            default:
+                return null; // see ConsoleBuiltins
+        }
+    }
+
+    public abstract static class JSAddTaintNode extends JSBuiltinNode {
+
+        private static final TruffleString DEFAULT_TAINT = constant("DEFAULT_TAINT");
+
+        // TODO is this indirection necessary (extending JSBuiltinNode),
+        //  or can I use TSTaintNode.AddTaintNode directly?
+
+        public JSAddTaintNode(JSContext context, JSBuiltin builtin) {
+            super(context, builtin);
+        }
+
+        @Specialization
+        AbstractTruffleString addTainted(AbstractTruffleString value, Object taint,
+                                             @Cached TSTaintNodes.AddTaintNode addTaintNode) {
+            return addTaintNode.execute(value, JSGuards.isUndefined(taint) ? DEFAULT_TAINT : taint);
+        }
+    }
+
+    public abstract static class JSIsTaintedNode extends JSBuiltinNode {
+
+        public JSIsTaintedNode(JSContext context, JSBuiltin builtin) {
+            super(context, builtin);
+        }
+
+        @Specialization
+        boolean isTainted(AbstractTruffleString a,
+                          @Cached TSTaintNodes.IsTaintedNode isTaintedNode) {
+            return isTaintedNode.execute(a);
+        }
+    }
+
+    public abstract static class JSGetTaintNode extends JSBuiltinNode {
+
+        public JSGetTaintNode(JSContext context, JSBuiltin builtin) {
+            super(context, builtin);
+        }
+
+        @Specialization(guards = "isTaintedNode.execute(a)")
+        JSArrayObject getTaintTainted(AbstractTruffleString a,
+                          @Cached TSTaintNodes.GetTaintNode getTaintNode,
+                          @Cached TSTaintNodes.IsTaintedNode isTaintedNode) {
+            final Object[] taint = mapNull(getTaintNode.execute(a));
+
+            // TODO no clue which array implementation I should choose, there are sooooo many
```

Comment:
As said, there are so many different array implementations, is there a preferred one I should use?

Reply:
There is only one array implementation in JavaScript. However, that implementation can use several different internal storage strategies to optimize itself to that data each individual array actually needs to store. Fortunately the array implementation anyways changes the storage strategy when a new value requires that. `JSArray.createEmpty` is a good choice for an empty array. However, you shouldn't bother with implementing the array creation yourself. JavaScript has an extensive body of builtins such as an array constructor and [`Array#push`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/push). Graal.js already implements nodes for all of them that you can just reuse. You can just have an `ArrayLiteralNode` and a `JSArrayPushNode` as `@Child` in this node and use them to create and fill your array for you.

Reply:
Is it a problem to just use `JSArray#createZeroBasedObjectArray()` like I do? It seems to work quite nicely

Reply:
It should be ok.